crates/polars-time/src/replace.rs

```diff
@@ -0,0 +1,163 @@
+    };
+    let minute = if minute.len() == 1 {
+        if let Some(value) = minute.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.minute()
+        }
+    } else {
+        &minute.zip_with(&minute.is_not_null(), &ca.minute())?
+    };
+    let second = if second.len() == 1 {
+        if let Some(value) = second.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.second()
+        }
+    } else {
+        &second.zip_with(&second.is_not_null(), &ca.second())?
+    };
+    let microsecond = if microsecond.len() == 1 {
+        if let Some(value) = microsecond.get(0) {
+            &Int32Chunked::full("".into(), value, n)
+        } else {
+            &(ca.nanosecond() / 1000)
+        }
+    } else {
+        &microsecond.zip_with(&microsecond.is_not_null(), &(ca.nanosecond() / 1000))?
+    };
+
+    let mut out = DatetimeChunked::new_from_parts(
+        year,
+        month,
+        day,
+        hour,
+        minute,
+        second,
+        microsecond,
+        ambiguous,
+        &ca.time_unit(),
+        ca.time_zone().as_deref(),
+        ca.name().clone(),
+    )?;
+
+    // Ensure nulls are propagated.
+    if ca.has_nulls() {
+        let mask = &ca.is_not_null();
+        let null = &Int64Chunked::full_null(PlSmallStr::EMPTY, n);
+        out = out
+            .physical()
+            .zip_with(mask, null)?
+            .into_datetime(ca.time_unit(), ca.time_zone().clone());
+    }
+
+    Ok(out)
+}
+
+/// Replace specific time component of a `DateChunked` with a specified value.
+#[cfg(feature = "dtype-date")]
+pub fn replace_date(
+    ca: &DateChunked,
+    year: &Int32Chunked,
+    month: &Int8Chunked,
+    day: &Int8Chunked,
+) -> PolarsResult<DateChunked> {
+    let n = ca.len();
+
+    let year = if year.len() == 1 {
+        if let Some(value) = year.get(0) {
+            &Int32Chunked::full("".into(), value, n)
+        } else {
+            &ca.year()
+        }
+    } else {
+        &year.zip_with(&year.is_not_null(), &ca.year())?
+    };
+    let month = if month.len() == 1 {
+        if let Some(value) = month.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.month()
+        }
+    } else {
+        &month.zip_with(&month.is_not_null(), &ca.month())?
+    };
+    let day = if day.len() == 1 {
+        if let Some(value) = day.get(0) {
+            &Int8Chunked::full("".into(), value, n)
+        } else {
+            &ca.day()
+        }
+    } else {
+        &day.zip_with(&day.is_not_null(), &ca.day())?
+    };
+    let mut out = DateChunked::new_from_parts(year, month, day, ca.name().clone())?;
+
+    // Ensure nulls are propagated.
+    if ca.has_nulls() {
+        let mask = &ca.is_not_null();
+        let null = &Int32Chunked::full_null(PlSmallStr::EMPTY, n);
+        out = out.physical().zip_with(mask, null)?.into_date();
+    }
```

Comment:
@ritchie is there a more efficient way to zip validities here?

Reply:
Yes, use arrow arrays. They have bitand/or kernels.

Reply:
@MarcoGorelli I'm going to re-draft this briefly while I figure out how to do this part a bit more efficiently.